Integration module company

NathaelB opened the pull request:
Implementation of routes for company CRUD, for the moment all routes are protected via Keycloak Roles.

Tests have been carried out (15).

NathaelB committed:
feat: create first route for companies

Signed-off-by: nathaelbonnal <pro.nathaelbonnal@gmail.com>

NathaelB committed:
feat: Update company and professional models

This commit updates the company and professional models. It renames the `company.ts` file to `shared/models/company.ts` and moves the `professional.ts` file to `shared/models/professional_type.ts`. It also creates a new file `shared/models/professional.ts` with the necessary model definitions. These changes are made to improve the organization and structure of the codebase.

Note: The recent user commits and repository commits are not relevant for generating the commit message.
Signed-off-by: nathaelbonnal <pro.nathaelbonnal@gmail.com>

NathaelB committed:
feat: Add company index route with authorization and validation

The code changes in this commit add an index route for companies. It includes authorization using the `bouncer` middleware and validation using the `getCompaniesValidator`. This allows only authorized users to view the list of companies and ensures that the request parameters are validated before processing.

Signed-off-by: nathaelbonnal <pro.nathaelbonnal@gmail.com>

NathaelB committed:
feat: Add company index route with authorization and validation

Signed-off-by: nathaelbonnal <pro.nathaelbonnal@gmail.com>

NathaelB committed:
feat: Add ability to edit user

This commit adds the ability to edit a user. It introduces a new ability called `editUser` in the `main.ts` file under the `apps/shared/abilities` directory. This ability is registered with the Bouncer instance and allows users to edit their own profile. This feature enhances the user experience by enabling them to update their information.

Note: The recent user commits and repository commits are not relevant for generating the commit message.
Signed-off-by: nathaelbonnal <pro.nathaelbonnal@gmail.com>

NathaelB committed:
feat: Add migrations for identities, professional types, and professionals

This commit adds migrations for three new tables: identities, professional_types, and professionals. The migrations create the necessary schema for each table, including primary keys, foreign keys, and timestamps. These tables are essential for storing user identity information, professional types, and professional details. This commit improves the database structure and prepares for future functionality.

Note: The recent user commits and repository commits are not relevant for generating the commit message.
Signed-off-by: nathaelbonnal <pro.nathaelbonnal@gmail.com>

NathaelB committed:
feat: Update database configuration for test environment

This commit updates the database configuration for the test environment. It changes the `DB_DATABASE` value from `backend` to `placecare_api_test` in the `.env.test` file. This change ensures that the tests use the correct database for the test environment.

Signed-off-by: nathaelbonnal <pro.nathaelbonnal@gmail.com>

NathaelB committed:
feat: Add show and store methods to CompaniesController

This commit adds the `show` and `store` methods to the `CompaniesController` class. The `show` method retrieves an individual record based on the provided ID, while the `store` method handles form submission for the create action. These changes enhance the functionality of the CompaniesController and improve the overall user experience.

Note: The recent user commits and repository commits are not relevant for generating the commit message.
Signed-off-by: nathaelbonnal <pro.nathaelbonnal@gmail.com>

NathaelB committed:
feat: Update adonisrc.ts to import commands and providers

The code changes in this commit update the `adonisrc.ts` file to import additional commands and providers. This enhances the functionality of the AdonisJS application by including commands from the `@adonisjs/core`, `@adonisjs/lucid`, and `@adonisjs/bouncer` packages. It also adds providers for CORS, database, authentication, and bouncer. These changes improve the overall capabilities of the application.

Signed-off-by: nathaelbonnal <pro.nathaelbonnal@gmail.com>

NathaelB committed:
feat: Add update and delete routes for companies

Signed-off-by: nathaelbonnal <pro.nathaelbonnal@gmail.com>

NathaelB committed:
feat: Update database configuration for test environment

This commit updates the database configuration for the test environment. It changes the `DB_DATABASE` value from `backend` to `placecare_api_test` in the `.env.test` file. This change ensures that the tests use the correct database for the test environment.

Signed-off-by: nathaelbonnal <pro.nathaelbonnal@gmail.com>

NathaelB committed:
feat: Add AddressService and ContactService classes

This commit adds the AddressService and ContactService classes to the shared/services directory. The AddressService class provides a create method for creating new addresses, while the ContactService class provides a create method for creating new contacts. These services enhance the functionality of the application by encapsulating the logic for creating addresses and contacts.

Note: The recent user commits and repository commits are not relevant for generating the commit message.
Signed-off-by: nathaelbonnal <pro.nathaelbonnal@gmail.com>

NathaelB committed:
feat: Add error handling for E_ROW_NOT_FOUND in HttpExceptionHandler

This commit adds error handling for the `E_ROW_NOT_FOUND` exception in the `HttpExceptionHandler` class. When this exception is thrown, the code now sends a 404 response with the error message, code, and status. This improves the user experience by providing a clear response when a requested resource is not found.

Signed-off-by: nathaelbonnal <pro.nathaelbonnal@gmail.com>

NathaelB committed:
feat: Add migration for creating company_professionals table

This commit adds a migration file for creating the `company_professionals` table in the database. The table has columns for `company_id` and `professional_id`, both of which are foreign keys referencing the `companies` and `professionals` tables respectively. The migration also includes a unique constraint on the combination of `company_id` and `professional_id`. This migration improves the database structure by establishing the relationship between companies and professionals.

Signed-off-by: nathaelbonnal <pro.nathaelbonnal@gmail.com>

NathaelB committed:
feat: Add tests for creating, updating, and deleting companies

This commit adds tests for creating, updating, and deleting companies in the `create.spec.ts`, `update.spec.ts`, and `delete.spec.ts` files respectively. The tests cover various scenarios, including successful operations, unauthorized access, and non-existing resources. These tests ensure the correctness and reliability of the company-related functionality in the application.

Signed-off-by: nathaelbonnal <pro.nathaelbonnal@gmail.com>

NathaelB committed:
feat: Add company seeder for Hopital de Montpellier and Hopital De Paris

This commit adds a new seeder file, `company_seeder.ts`, which creates two company records: "Hopital de Montpellier" and "[delete] Hopital De Paris". The seeder also creates associated address and contact records for each company. This addition enhances the database with sample data for testing and development purposes.

Note: The recent user commits and repository commits are not relevant for generating the commit message.
Signed-off-by: nathaelbonnal <pro.nathaelbonnal@gmail.com>

NathaelB committed:
feat: Update CompaniesController to include update method

This commit updates the CompaniesController class to include the update method. The update method handles form submission for the edit action, allowing users to update company information. This addition enhances the functionality of the CompaniesController and improves the overall user experience.

Signed-off-by: nathaelbonnal <pro.nathaelbonnal@gmail.com>

NathaelB committed:
feat: Update logger configuration to enable/disable based on environment

This commit updates the logger configuration in the `logger.ts` file. The `enabled` property of the `app` logger is now set based on the value of the `NODE_ENV` environment variable. If the `NODE_ENV` is not equal to 'test', the logger will be enabled. This change ensures that the logger is only active in non-test environments, improving performance and reducing unnecessary logging in tests.

Signed-off-by: nathaelbonnal <pro.nathaelbonnal@gmail.com>

LeadcodeDev commented on apps/company/services/company_service.ts:
```diff
@@ -1,7 +1,65 @@
+      addressId: address.id,
+      contactId: contact.id,
+    })
+    await company.load('address')
```
You can use `Promise.all()` instead

LeadcodeDev commented on tests/functional/companies/delete.spec.ts:
```diff
@@ -0,0 +1,57 @@
+    const professional = await Professional.firstOrFail()
+
+    const response = await client
+      .delete('/v1/companies/9890150790490362881976491')
```
When you would test, you should to make factory instead seeder.
Then call your factory like `const model = await Factory.make(…)` to create model. instance.

So you can get `model.id` to replace `9890150790490362881976491`

NathaelB replied:
For the moment I've made a simple seeder, maybe in time we'll do something cleaner with Factory.

LeadcodeDev commented on tests/functional/companies/update.spec.ts:
```diff
@@ -0,0 +1,75 @@
+    const professional = await Professional.firstOrFail()
+
+    const response = await client
+      .put('/v1/companies/9890153540490362881974927')
```
Same as below

NathaelB committed:
refactor: Use Promise.all for loading address and contact in CompanyService

This commit refactors the `CompanyService` class in the `company_service.ts` file. Instead of using separate `await` statements to load the address and contact, this change uses `Promise.all` to load them concurrently. This improves the efficiency of the code by reducing the overall execution time.

Signed-off-by: nathaelbonnal <pro.nathaelbonnal@gmail.com>
